crates/autopilot/src/infra/persistence/mod.rs

```diff
@@ -135,12 +137,207 @@ impl Persistence {
+
+            let mut prices = HashMap::new();
+            for price in db_prices {
+                let token = eth::H160(price.token.0).into();
+                let price = big_decimal_to_u256(&price.price)
+                    .ok_or(domain::auction::InvalidPrice)
+                    .and_then(|p| domain::auction::Price::new(p.into()))
+                    .map_err(error::Auction::Price)?;
+                prices.insert(token, price);
+            }
+            prices
+        };
+
+        let fee_policies = {
+            let orders = database::auction_orders::fetch(&mut ex, auction_id)
+                .await
+                .context("fetch auction orders")?
+                .ok_or(error::Auction::MissingOrders)?
+                .into_iter()
+                .map(|order| domain::OrderUid(order.0));
+
+            let mut fee_policies = HashMap::new();
+            for order in orders {
+                fee_policies.insert(order, vec![]);
+                if database::orders::read_order(&mut ex, &ByteArray(order.0))
+                    .await
+                    .context("fetch order")?
+                    .is_some()
+                {
+                    let quote = database::orders::read_quote(&mut ex, &ByteArray(order.0))
+                        .await
+                        .context("fetch quote")?
+                        .map(dto::quote::into_domain)
+                        .transpose()
+                        .map_err(|_| error::Auction::DbConversion("quote overflow"))?
+                        .ok_or(error::Auction::MissingQuote)?;
+
+                    let policies =
+                        database::fee_policies::fetch(&mut ex, auction_id, ByteArray(order.0))
+                            .await
+                            .context("fetch fee policies")?;
+
+                    for policy in policies {
+                        let policy = dto::fee_policy::into_domain(policy, &quote)?;
+                        fee_policies.entry(order).and_modify(|policies| {
+                            policies.push(policy);
+                        });
+                    }
+                }
+            }
+            fee_policies
+        };
+
+        Ok(domain::settlement::Auction {
+            id: auction_id,
+            prices,
+            deadline,
+            fee_policies,
+        })
+    }
+
+    /// Get winning competition solution.
+    pub async fn get_competition_solution(
+        &self,
+        auction_id: domain::auction::Id,
+    ) -> Result<domain::competition::Solution, error::Solution> {
+        let mut ex = self.postgres.pool.begin().await.context("begin")?;
+
+        let (winner, score) = {
+            let scores = database::settlement_scores::fetch(&mut ex, auction_id)
+                .await
+                .context("fetch scores")?
+                // if score is missing, no competition / auction exist for this auction_id
+                .ok_or(error::Solution::MissingScore)?;
```

Comment:
Here also it's not really a missing score, but rather a missing solution (it's hard for the callsite to reason that a particular missing property like "score" should imply there has never been a valid solution for this auction).